src/uct/ib/dc/dc_mlx5.c

```diff
@@ -1316,3 +1322,89 @@ uct_dc_mlx5_query_tl_devices(uct_md_h md, uct_tl_device_resource_t **tl_devices_
 UCT_TL_DEFINE(&uct_ib_component, dc_mlx5, uct_dc_mlx5_query_tl_devices,
               uct_dc_mlx5_iface_t, "DC_MLX5_", uct_dc_mlx5_iface_config_table,
               uct_dc_mlx5_iface_config_t);
+
+static void
+uct_dc_mlx5_dci_keepalive_handle_failure(uct_dc_mlx5_iface_t *iface,
+                                         struct mlx5_cqe64 *cqe,
+                                         uint8_t dci,
+                                         ucs_status_t ep_status)
+{
+    uct_ib_iface_t *ib_iface = &iface->super.super.super;
+    uct_dc_mlx5_ep_t *ep;
+    uct_rc_txqp_t *txqp;
+    uct_ib_mlx5_txwq_t *txwq;
+    uct_rc_iface_send_op_t *op;
+    uint16_t hw_ci;
+    ucs_queue_elem_t *elem;
+    ucs_status_t status;
+    ucs_log_level_t log_lvl;
+
+    ucs_assert(dci == iface->tx.ndci);
+
+    hw_ci  = ntohs(cqe->wqe_counter);
+    txqp   = &iface->tx.dcis[dci].txqp;
+    elem   = ucs_queue_pull(&txqp->outstanding);
+    if (elem == NULL) {
+        /* outstanding list is empty, just exit */
+        goto reset_dci;
+    }
+
+    op = ucs_container_of(elem, uct_rc_iface_send_op_t, queue);
+    if (hw_ci != op->sn) {
+        goto put_op;
+    }
```

Comment:
how could this happen?

Reply:
but then the queue would be empty, no? we check this at line 1341

Reply:
i think in this case still (hw_ci == op->sn) because we haven't changed anything in the keeaplive ep tx queue when we destroyed the ep.
on the other hand this indicates a new problem: we can be using op->ep after it's already destroyed
so when destroying dc ep which has a posted keepalive, we need to find it and remove it, or at mark it as invalid